Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/TrackParameterSmearing.cpp

```diff
@@ -0,0 +1,145 @@
+
+#include "ActsExamples/TruthTracking/TrackParameterSmearing.hpp"
+
+#include "Acts/Seeding/EstimateTrackParamsFromSeed.hpp"
+#include "ActsExamples/EventData/Track.hpp"
+#include "ActsExamples/Framework/AlgorithmContext.hpp"
+#include "ActsExamples/Framework/RandomNumbers.hpp"
+
+#include <random>
+#include <stdexcept>
+#include <utility>
+
+namespace ActsExamples {
+
+TrackParameterSmearing::TrackParameterSmearing(const Config& config,
+                                               Acts::Logging::Level level)
+    : IAlgorithm("TrackParameterSmearing", level), m_cfg(config) {
+  if (m_cfg.inputTrackParameters.empty()) {
+    throw std::invalid_argument("Missing input track parameters collection");
+  }
+  if (m_cfg.outputTrackParameters.empty()) {
+    throw std::invalid_argument("Missing output track parameters collection");
+  }
+  if (m_cfg.randomNumbers == nullptr) {
+    throw std::invalid_argument("Missing random numbers tool");
+  }
+
+  if (m_cfg.particleHypothesis) {
+    ACTS_INFO("Override truth particle hypothesis with "
+              << *m_cfg.particleHypothesis);
+  }
+
+  m_inputTrackParameters.initialize(m_cfg.inputTrackParameters);
+  m_outputTrackParameters.initialize(m_cfg.outputTrackParameters);
+}
+
+ProcessCode TrackParameterSmearing::execute(const AlgorithmContext& ctx) const {
+  // setup input and output containers
+  const auto& inputTrackParametersContainer = m_inputTrackParameters(ctx);
+  TrackParametersContainer outputTrackParametersContainer;
+  outputTrackParametersContainer.reserve(inputTrackParametersContainer.size());
+
+  // setup random number generator and standard gaussian
+  auto rng = m_cfg.randomNumbers->spawnGenerator(ctx);
+  std::normal_distribution<double> stdNormal(0.0, 1.0);
+
+  for (const auto& inputTrackParameters : inputTrackParametersContainer) {
+    const auto time = inputTrackParameters.time();
+    const auto phi = inputTrackParameters.phi();
+    const auto theta = inputTrackParameters.theta();
+    const auto pt = inputTrackParameters.transverseMomentum();
+    const auto qOverP = inputTrackParameters.qOverP();
+    const auto particleHypothesis = m_cfg.particleHypothesis.value_or(
+        inputTrackParameters.particleHypothesis());
+
+    // compute momentum-dependent resolutions
+    const double sigmaLoc0 =
+        m_cfg.sigmaLoc0 +
+        m_cfg.sigmaLoc0PtA * std::exp(-1.0 * std::abs(m_cfg.sigmaLoc0PtB) * pt);
+    const double sigmaLoc1 =
+        m_cfg.sigmaLoc1 +
+        m_cfg.sigmaLoc1PtA * std::exp(-1.0 * std::abs(m_cfg.sigmaLoc1PtB) * pt);
+    // shortcuts for other resolutions
+    const double sigmaTime = m_cfg.sigmaTime;
+    const double sigmaPhi = m_cfg.sigmaPhi;
+    const double sigmaTheta = m_cfg.sigmaTheta;
+    const double sigmaQOverP =
+        std::sqrt(std::pow(m_cfg.sigmaPtRel * qOverP, 2) +
+                  std::pow(sigmaTheta * (qOverP * std::tan(theta)), 2));
+
+    Acts::BoundVector params = Acts::BoundVector::Zero();
```

Comment:
Again mega pedantic; you overwrite all components so probably no need to initialize. Shouldn't matter performance wise anyway.

Reply:
Yeah this is not pretty. I don't see a great way around this. Otherwise we have to create more temporary variables. In case somebody accidentally drops a variable it will be initialized which is good I think